### DISCUSSION TITLE
g2o error on adding "frame_publsiher.h" in custom executable

### DISCUSSION THREAD

**Discussion by Sang-Beom-Woo** [stella-cv/stella_vslam#205] [Q&A] [answered]
Hello, I am making coustom ROS2 node in openvslam_ros package("openvslam_ros_mod.cpp  and hpp") to get the current frame image and send it as image topic. The problem occur when adding "openvlsam/publish/ frame_publisher.h" in to my wrapper code. The compliler outputs following error message when compiling the preprocessor;

In file included from /usr/local/include/g2o/core/base_fixed_sized_edge.h:292,
                 from /usr/local/include/g2o/types/sba/edge_project_psi2uv.h:30,
                 from /usr/local/include/g2o/types/sba/types_six_dof_expmap.h:38,
                 from /usr/local/include/openvslam/util/converter.h:8,
                 from /usr/local/include/openvslam/data/frame.h:6,
                 from /usr/local/include/openvslam/tracking_module.h:5,
                 from /usr/local/include/openvslam/publish/frame_publisher.h:5,
                 from /home/robotis/colcon_ws/src/openvslam_ros/src/openvslam_ros_mod.cpp:4:
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp: In instantiation of ‘void g2o::BaseFixedSizedEdge<D, E, VertexTypes>::linearizeOplusN() [with int N = 0; int D = 3; E = Eigen::Matrix<double, 3, 1>; VertexTypes = {g2o::VertexSE3Expmap}]’:
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:192:43:   required from ‘void g2o::BaseFixedSizedEdge<D, E, VertexTypes>::linearizeOplusNs(std::index_sequence<Ints ...>) [with long unsigned int ...Ints = {0}; int D = 3; E = Eigen::Matrix<double, 3, 1>; VertexTypes = {g2o::VertexSE3Expmap}; std::index_sequence<Ints ...> = std::integer_sequence<long unsigned int, 0>]’
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:200:3:   required from ‘void g2o::BaseFixedSizedEdge<D, E, VertexTypes>::linearizeOplus() [with int D = 3; E = Eigen::Matrix<double, 3, 1>; VertexTypes = {g2o::VertexSE3Expmap}]’
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:197:6:   required from here
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:166:14: error: ‘FixedArray’ {aka ‘class ceres::internal::FixedArray<double, 6>’} has no member named ‘fill’
  166 |   add_vertex.fill(0.);
      |   ~~~~~~~~~~~^~~~
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:173:30: error: ‘FixedArray’ {aka ‘class ceres::internal::FixedArray<double, 6>’} has no member named ‘data’
  173 |     vertex->oplus(add_vertex.data());
      |                   ~~~~~~~~~~~^~~~
/usr/local/include/g2o/core/base_fixed_sized_edge.hpp:179:30: error: ‘FixedArray’ {aka ‘class ceres::internal::FixedArray<double, 6>’} has no member named ‘data’
  179 |     vertex->oplus(add_vertex.data());
      |                   ~~~~~~~~~~~^~~~

one wired thing is that, src/pangolin_viewer/viewer.cc in openvslam package compiled ok without any error messages, though it also declare "openvlsam/publish/ frame_publisher.h" as i do in the preprocessor.

The things I tried so far is shown below and none of them works.

1. compile g2o and openvslam package with c++14 option and sudo make install

2. compile g2o with master brach instead of "9b41a4ea5ade8e1250b9c1b279f3a9c098811b5a" branch recommend by community

Is anyone have same issue and know how to solve it?

**Comment by ymd-stella** [comment, marked as answer]
Try to install `ros-${ROS_DISTRO}-libg2o` by apt.

**Reply by ymd-stella** [reply]
I don't know if it will help in your case too, but in my environment, https://github.com/OpenVSLAM-Community/openvslam/pull/222 was helpful.

**Reply by Sang-Beom-Woo** [reply]
Thank you! It worked somehow with `apt-get install ros-${ROS_DISTRO}-libg2o`

**Reply by mgrallos** [reply]
> Try to install `ros-${ROS_DISTRO}-libg2o` by apt.

Thank you that also work for me !
